File: libgomp/testsuite/libgomp.fortran/taskloop3.f90
! { dg-do run }
! { dg-options "-O2" }

  integer, save :: g
  integer :: i
  !$omp parallel
  !$omp single
    if (f1 (74) .ne. 63 + 4) call abort
    g = 77
    call f2
    !$omp taskwait
    if (g .ne. 63 + 9) call abort
    if (f3 (7_8, 11_8, 2_8) .ne. 11 * 7 + 13) call abort
    if (f4 (0_8, 31_8, 16_8, 46_8, 1_8, 2_8, 73) .ne. 32 + 5 * 48 &
&       + 11 * 31 + 17 * 46) call abort
  !$omp end single
  !$omp end parallel
contains
  function f1 (y)
    integer, intent(in) :: y
    integer :: i, f1, x
    x = y
    !$omp taskloop firstprivate(x)lastprivate(x)
    do i = 0, 63
      if (x .ne. 74) call abort
      if (i .eq. 63) then
        x = i + 4
      end if
    end do
    f1 = x
  end function f1
  subroutine f2 ()
    integer :: i
    !$omp taskloop firstprivate(g)lastprivate(g)nogroup
    do i = 0, 63
      if (g .ne. 77) call abort
      if (i .eq. 63) then
        g = i + 9
      end if
    end do
  end subroutine f2
  function f3 (a, b, c)
    integer(kind=8), intent(in) :: a, b, c
    integer(kind=8) :: i, f3
    integer :: l
    !$omp taskloop default(none) lastprivate (i, l)
    do i = a, b, c
      l = i
    end do
    !$omp end taskloop
    f3 = l * 7 + i
  end function f3
  function f4 (a, b, c, d, e, f, m)
    integer(kind=8), intent(in) :: a, b, c, d, e, f
    integer(kind=8) :: i, j, f4
    integer, intent(in) :: m
    integer :: l, k
    k = m
    !$omp taskloop default (none) collapse (2) firstprivate (k) &
    !$omp & lastprivate (i, j, k, l)
    do i = a, b, e
      do j = c, d, f
        if (k .ne. 73) call abort
        if (i .eq. 31 .and. j .eq. 46) then
          k = i
        end if
        l = j
      end do
    end do
    f4 = i + 5 * j + 11 * k + 17 * l
  end function f4
end

File: libgomp/testsuite/libgomp.c/taskloop-3.c
/* { dg-do run } */
/* { dg-options "-O2 -std=c99" } */

int g;
int a[1024];

__attribute__((noinline, noclone)) int
f1 (int x)
{
  #pragma omp taskloop firstprivate (x) lastprivate (x)
  for (int i = 0; i < 64; i++)
    {
      if (x != 74)
	__builtin_abort ();
      if (i == 63)
	x = i + 4;
    }
  return x;
}

__attribute__((noinline, noclone)) void
f2 (void)
{
  #pragma omp taskloop firstprivate (g) lastprivate (g) nogroup
  for (int i = 0; i < 64; i++)
    {
      if (g != 77)
	__builtin_abort ();
      if (i == 63)
	g = i + 9;
    }
}

__attribute__((noinline, noclone)) long long
f3 (long long a, long long b, long long c)
{
  long long i;
  int l;
  #pragma omp taskloop default (none) lastprivate (i, l)
  for (i = a; i < b; i += c)
    l = i;
  return l * 7 + i;
}

__attribute__((noinline, noclone)) long long
f4 (long long a, long long b, long long c, long long d,
    long long e, long long f, int k)
{
  long long i, j;
  int l;
  #pragma omp taskloop default (none) collapse(2) \
	      firstprivate (k) lastprivate (i, j, k, l)
  for (i = a; i < b; i += e)
    for (j = c; j < d; j += f)
      {
	if (k != 73)
	  __builtin_abort ();
	if (i == 31 && j == 46)
	  k = i;
	l = j;
      }
  return i + 5 * j + 11 * k + 17 * l;
}

int
main ()
{
  #pragma omp parallel
    #pragma omp single
      {
	if (f1 (74) != 63 + 4)
	  __builtin_abort ();
	g = 77;
	f2 ();
	#pragma omp taskwait
	if (g != 63 + 9)
	  __builtin_abort ();
	if (f3 (7, 12, 2) != 11 * 7 + 13)
	  __builtin_abort ();
	if (f4 (0, 32, 16, 48, 1, 2, 73) != 32 + 5 * 48 + 11 * 31 + 17 * 46)
	  __builtin_abort ();
      }
  return 0;
}
